asmcomp/selectgen.ml

```diff
@@ -548,78 +542,156 @@ method select_condition = function
    ports (e.g. Arm) can override this definition to store float values
    in pairs of integer registers. *)
 
-method regs_for tys = Reg.createv tys
-
-(* Buffering of instruction sequences *)
-
-val mutable instr_seq = dummy_instr
+let default_regs_for _self tys = Reg.createv tys
 
-method insert_debug _env desc dbg arg res =
-  instr_seq <- instr_cons_debug desc arg res dbg instr_seq
+let insert_debug self _env desc dbg arg res =
+  self.instr_seq <- instr_cons_debug desc arg res dbg self.instr_seq
 
-method insert _env desc arg res =
-  instr_seq <- instr_cons desc arg res instr_seq
+let insert self _env desc arg res =
+  self.instr_seq <- instr_cons desc arg res self.instr_seq
 
-method extract_onto o =
+let extract_onto self o =
   let rec extract res i =
     if i == dummy_instr
       then res
       else extract {i with next = res} i.next in
-    extract o instr_seq
+    extract o self.instr_seq
 
-method extract =
-  self#extract_onto (end_instr ())
+let extract self =
+  extract_onto self (end_instr ())
 
 (* Insert a sequence of moves from one pseudoreg set to another. *)
 
-method insert_move env src dst =
+let insert_move self env src dst =
   if src.stamp <> dst.stamp then
-    self#insert env (Iop Imove) [|src|] [|dst|]
+    insert self env (Iop Imove) [|src|] [|dst|]
 
-method insert_moves env src dst =
+let insert_moves self env src dst =
   for i = 0 to Stdlib.Int.min (Array.length src) (Array.length dst) - 1 do
-    self#insert_move env src.(i) dst.(i)
+    insert_move self env src.(i) dst.(i)
   done
 
 (* Insert moves and stack offsets for function arguments and results *)
 
-method insert_move_args env arg loc stacksize =
+let insert_move_args self env arg loc stacksize =
   if stacksize <> 0 then begin
-    self#insert env (Iop(Istackoffset stacksize)) [||] [||]
+    insert self env (Iop(Istackoffset stacksize)) [||] [||]
   end;
-  self#insert_moves env arg loc
+  insert_moves self env arg loc
 
-method insert_move_results env loc res stacksize =
+let insert_move_results self env loc res stacksize =
   if stacksize <> 0 then begin
-    self#insert env (Iop(Istackoffset(-stacksize))) [||] [||]
+    insert self env (Iop(Istackoffset(-stacksize))) [||] [||]
   end;
-  self#insert_moves env loc res
+  insert_moves self env loc res
 
 (* Add an Iop opcode. Can be overridden by processor description
    to insert moves before and after the operation, i.e. for two-address
    instructions, or instructions using dedicated registers. *)
 
-method insert_op_debug env op dbg rs rd =
-  self#insert_debug env (Iop op) dbg rs rd;
+let default_insert_op_debug self env op dbg rs rd =
+  insert_debug self env (Iop op) dbg rs rd;
   rd
 
-method insert_op env op rs rd =
-  self#insert_op_debug env op Debuginfo.none rs rd
+let default_insert_op self env op rs rd =
+  self.insert_op_debug self env op Debuginfo.none rs rd
```

Comment:
I don't like the fact that some operations are called like `insert_debug self env ...`, while others are called like `self.insert_op_debug self env ...`. In the object version you don't have to know whether a method is meant for overriding or not when you call it, and this makes the code more regular. (It also makes it easier to decide to override more functions later, etc.)

I may be missing something, but I think that you could either:
1. define `let <foo> self = self.<foo> self` for all fields `<foo>`, and consistently use `<foo> self`
2. or consistently put all self-taking functions in the record, and consistently use `self.<foo> self`

((1) works well with your choice of defining `let default_<foo> self = ...` functions for functions meant to be overriden.)

Reply:
I was making a distinction in my mind between functions that can be overrided and functions that can't. The `default_*` naming scheme was actually so that I never called an overridable function directly by mistake.
But I agree that your suggestion (1) makes sense, I'll update the code to use that.

I think that putting all self-taking functions is the record would be a shame. In particular for the recursive functions, going through the record would turn direct calls into indirect calls, which has a sizeable impact on performance. If I get enough complaints that it makes it harder to assess whether the code has changed or not, I might reluctantly go back to a more mechanical transformation. But note that some of these functions were only exported because of technical details (see comments in `selectgen.mli`), so this looked like a good opportunity to fix that.